Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/Capture.cpp

```diff
@@ -0,0 +1,844 @@
+        {
+            goto done;
+        }
+    }
+
+
+    hr = m_pEngine->StartPreview();
+    if (!m_fPowerRequestSet && m_hpwrRequest != INVALID_HANDLE_VALUE)
+    {
+        // NOTE:  By calling this, on SOC systems (AOAC enabled), we're asking the system to not go
+        // into sleep/connected standby while we're streaming.  However, since we don't want to block
+        // the device from ever entering connected standby/sleep, we're going to latch ourselves to
+        // the monitor on/off notification (RegisterPowerSettingNotification(GUID_MONITOR_POWER_ON)).
+        // On SOC systems, this notification will fire when the user decides to put the device in
+        // connected standby mode--we can trap this, turn off our media streams and clear this
+        // power set request to allow the device to go into the lower power state.
+        m_fPowerRequestSet = (TRUE == PowerSetRequest(m_hpwrRequest, PowerRequestExecutionRequired));
+    }
+done:
+    SafeRelease(&pSink);
+    SafeRelease(&pMediaType);
+    SafeRelease(&pMediaType2);
+    SafeRelease(&pSource);
+
+    return hr;
+}
+
+HRESULT CaptureManager::StopPreview()
+{
+    HRESULT hr = S_OK;
+
+    if (m_pEngine == NULL)
+    {
+        return MF_E_NOT_INITIALIZED;
+    }
+
+    if (!m_bPreviewing)
+    {
+        return S_OK;
+    }
+    hr = m_pEngine->StopPreview();
+    if (FAILED(hr))
+    {
+        goto done;
+    }
+    WaitForResult();
+
+    if (m_fPowerRequestSet && m_hpwrRequest != INVALID_HANDLE_VALUE)
+    {
+        PowerClearRequest(m_hpwrRequest, PowerRequestExecutionRequired);
+        m_fPowerRequestSet = false;
+    }
+done:
+    return hr;
+}
+
+// Helper function to get the frame size from a video media type.
+inline HRESULT GetFrameSize(IMFMediaType *pType, UINT32 *pWidth, UINT32 *pHeight)
+{    return MFGetAttributeSize(pType, MF_MT_FRAME_SIZE, pWidth, pHeight);}
+
+// Helper function to get the frame rate from a video media type.
+inline HRESULT GetFrameRate(
+    IMFMediaType *pType, 
+    UINT32 *pNumerator, 
+    UINT32 *pDenominator
+    )
+{
+    return MFGetAttributeRatio(
+        pType, 
+        MF_MT_FRAME_RATE, 
+        pNumerator, 
+        pDenominator
+        );
+}
+
+
+HRESULT GetEncodingBitrate(IMFMediaType *pMediaType, UINT32 *uiEncodingBitrate)
+{
+    UINT32 uiWidth;
+    UINT32 uiHeight;
+    float uiBitrate;
+    UINT32 uiFrameRateNum;
+    UINT32 uiFrameRateDenom;
+
+    HRESULT hr = GetFrameSize(pMediaType, &uiWidth, &uiHeight);
+    if (FAILED(hr))
+    {
+        goto done;
+    }
+
+    hr = GetFrameRate(pMediaType, &uiFrameRateNum, &uiFrameRateDenom);
+    if (FAILED(hr))
+    {
+        goto done;
+    }
+
+    uiBitrate = uiWidth / 3.0f * uiHeight * uiFrameRateNum / uiFrameRateDenom;
+    
+    *uiEncodingBitrate = (UINT32) uiBitrate;
+
+done:
+
+    return hr;
+}
+
+HRESULT ConfigureVideoEncoding(IMFCaptureSource *pSource, IMFCaptureRecordSink *pRecord, REFGUID guidEncodingType)
+{
+    IMFMediaType *pMediaType = NULL;
```

Comment:
>IMFMediaType [](http://example.com/codeflow?start=4&length=12)

I think using WIL auto pointers and return logic for external samples, it can remove issues with leaks and make the code smaller

Reply:
Removed the lingering raw pointers in capture